### PR TITLE
[dynamo] add lazy IteratorVariable implementations for map and zip

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131725
* #131716
* __->__ #131413

Fixes https://github.com/pytorch/pytorch/issues/130750.

Repro of lazy/eager `map` discrepancy without `islice`:
```python
    def fn(a, b):
        y = 1

        def f(x):
            nonlocal y
            y += 1
            return x

        l = list(zip([a, b], map(f, [1, 2, 3, 4])))
        return a + y
```

The major change is that we implement `MapVariable` and `ZipVariable` based on `IteratorVariable`. Before, `map` and `zip` were being traced by immediately unpacking the result as a `TupleVariable`, which is wrong in cases such as the example above.

`MapVariable`s are not allowed to be unpacked while `ZipVariable`s can only be unpacked if all of its iterables can also be unpacked.

We also add new `[has_]force_unpack_var_sequence` methods to `VariableTracker` for the case where it is safe to unpack the entire sequence lazily, e.g., when building a list from a map (i.e. `list(map(f, ...))`).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/iter.py [hunk 1]
```diff
@@ -263,10 +289,155 @@ def next_variable(self, tx):
...
+        self.iterables = iterables
+        self.index = 0
+        self.strict = strict
+
+    def python_type(self):
+        return zip
+
+    def has_unpack_var_sequence(self, tx) -> bool:
+        return all(
+            isinstance(it, list) or it.has_unpack_var_sequence(tx)
```
Comment (anijain2305) [inline torch/_dynamo/variables/iter.py:324]: Why only list?

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/iter.py (hunk 1 above)
Comment (williamwen42) [reply torch/_dynamo/variables/iter.py:324]: `it` is either `VariableTracker` or `List[VariableTracker]` (from `unpack_var_sequence` called from the `Map/ZipVariable` callsite).

**Pull request closed without merging**